src/main/java/appeng/crafting/v2/FastToIterateArrayBasedMultiMap.java

```diff
@@ -0,0 +1,258 @@
+package appeng.crafting.v2;
+
+import java.lang.reflect.Array;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A simple map implementation that uses an array for the keys and an array of arrays for the values. This map is only
+ * intended for a very small number of keys and values, in a scenario where the map is initialized once and afterward
+ * only queried. All operations are slow, except for manually iterating over {@link #keyArray()} and
+ * {@link #valuesArrayAt(int)}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public final class FastToIterateArrayBasedMultiMap<K, V> {
```

Comment:
Was there nothing of use in fastutils?

Reply:
I think you could hack something together using `Object2ObjectArrayMap`. Maybe I dismissed that too early